File: frontend/src/types/annotation.ts
import type { Point3D, Quaternion, Dimensions3D } from './common'

export interface AnnotationUpdateRequest {
  center?:     Point3D
  dimensions?: Dimensions3D
  rotation?:   Quaternion
  attributes?: string[]
}

// GET /api/v1/categories レスポンスの各エントリ（CategoryResponse に対応）
export interface Category {
  token:       string
  name:        string
  description: string | null
}

// GET /api/v1/annotations レスポンスの各エントリ（AnnotationResponse に対応）
export interface Attribute {
  token:       string
  name:        string
  description: string | null
}

export interface Visibility {
  token:       string
  level:       string
  description: string | null
}

export interface Annotation {
  token:            string
  sample_token:     string
  instance_token:   string
  translation:      number[]   // [x, y, z] グローバル座標
  rotation:         number[]   // [w, x, y, z] クォータニオン
  size:             number[]   // [width, length, height]
  prev:             string | null
  next:             string | null
  num_lidar_pts:    number
  num_radar_pts:    number
  visibility_token: string | null
  category_token:   string
  attributes:       Attribute[]
  visibility:       Visibility | null
}

// GET /api/v1/instances/{token}/annotations レスポンス（InstanceAnnotationResponse に対応）
export interface InstanceAnnotation extends Annotation {
  timestamp: number
}

// GET /api/v1/instances/{token} レスポンス（InstanceResponse に対応）
export interface Instance {
  token:                  string
  category_token:         string
  category_name:          string
  nbr_annotations:        number
  first_annotation_token: string | null
  last_annotation_token:  string | null
}

File: backend/app/schemas/annotation.py
from pydantic import BaseModel, ConfigDict


class AttributeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    token: str
    name: str
    description: str | None


class CategoryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    token: str
    name: str
    description: str | None


class VisibilityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    token: str
    level: str
    description: str | None


class AnnotationResponse(BaseModel):
    token: str
    sample_token: str
    instance_token: str
    translation: list[float]   # [x, y, z] 中心座標（グローバル座標）
    rotation: list[float]      # [w, x, y, z] クォータニオン
    size: list[float]          # [width, length, height]
    prev: str | None
    next: str | None
    num_lidar_pts: int
    num_radar_pts: int
    visibility_token: str | None
    # instance 経由で展開
    category_token: str
    attributes: list[AttributeResponse]
    visibility: VisibilityResponse | None


class SampleInstanceResponse(BaseModel):
    instance_token: str
    category_name: str
    nbr_annotations: int


class InstanceAnnotationResponse(AnnotationResponse):
    """instances/{token}/annotations 用: AnnotationResponse + timestamp"""
    timestamp: int


class InstanceResponse(BaseModel):
    """GET /instances および GET /instances/{token} のレスポンス"""
    token:                  str
    category_token:         str
    category_name:          str          # Instance.category.name リレーションシップから
    nbr_annotations:        int
    first_annotation_token: str | None
    last_annotation_token:  str | None

# ------ Update ------

class AnnotationUpdate(BaseModel):
    """PATCH 用: 送ったフィールドだけ更新する"""
    translation: list[float] | None = None
    rotation: list[float] | None = None
    size: list[float] | None = None
    visibility_token: str | None = None
    attribute_tokens: list[str] | None = None

# ------ Create ------

class InstanceCreate(BaseModel):
    """新規 Instance 作成用ペイロード."""
    category_token: str


class AnnotationCreate(BaseModel):
    """POST /annotations 用: 新規 BBox 追加.

    instance_token と new_instance は排他的:
      - instance_token が指定されれば既存 Instance に追加
      - new_instance が指定されれば新規 InstanceEdit を作成して追加
    """
    sample_token:     str
    instance_token:   str | None = None
    new_instance:     InstanceCreate | None = None
    translation:      list[float]
    rotation:         list[float]
    size:             list[float]
    prev:             str | None = None
    next:             str | None = None
    visibility_token: str | None = None
    attribute_tokens: list[str] = []
